Nos/Models/NoteParser.swift

```diff
@@ -1,19 +1,20 @@
 import CoreData
 import Foundation
+import Logger
 import RegexBuilder
 
 /// This struct encapsulates the algorithms that parse notes and the mentions inside the note.
-enum NoteParser {
+struct NoteParser {
```

Comment:
@joshuatbrown @martindsq For some reason I can't find the comment about this change. Can you link it to me or explain why it was made? Each of the functions below are pure, and I find it a best practice to make pure functions `static` because it is a clear signal to the developer: "this function doesn't use self" If you want to collect a bunch of related pure functions together it makes sense to do so in a stateless `enum`, because you can't instantiate a stateless `enum`. This is a pattern we use across the codebase. The `struct` here allows us to instantiate a `NoteParser`, but why? NoteParser doesn't need to store anything in memory.

Reply:
The discussion that sparked this change is this one https://github.com/planetary-social/nos/pull/1043#discussion_r1569388861. At the time of adding NoteParser, changing it ti enum was also one of your suggestions https://github.com/planetary-social/nos/pull/281#discussion_r1181859468.

1. `enum` are for enumerations and that's it, if we use it in a different and unique way, that will confuse the clients of `NoteParser` because the signature will be semantically incongruent. 

2. I think the clients of NoteParser should not care about if the functions inside use `self` or not. That's specific to the implementation. The best practice here, taking into account Design Patterns, is to encapsulate a complex algorithm in an object and hide the complexity in an interface. NoteParser doesn't need to store anything in memory, but it could later, it could start having parameters the user could tune in the UI, or we could be adding some private vars to make the code easier to read. And that kind of change in the implementation of NoteParser won't affect all clients of NoteParser if it was already an object, but it will if it was a bunch of static functions before.

Reply:
Hm we are definitely in disagreement here. 

> enum are for enumerations and that's it

Using caseless enums is a fairly common pattern in Swift in my experience, in fact it's in the default swiftlint rules. If you create a struct without all static functions it will warn you to change it to a caseless enum. It's true, a caseless enum is not an enumeration which might be confusing the first time you see it, but you actually _can't misuse` it by instantiating one which is the point of its use.

Also I see from the linked comment that both you and @joshuatbrown prefer not to use static functions. Can you explain more why that is? Part of my development philosophy is to make types as strict as you can. For instance, I wouldn't want to pass a string to a function if you can instead use an enum. Similarly a static function is a stricter type of function, it is not attached to a class or value so (barring mutation of global variables or other no-nos) you know that it will clean up all its memory when the function ends. 

If I want to parse a note and I'm seeing in autocomplete that I can just call `NoteParser.parse(...)` then I know I can just autocomplete that and I'm good to go. If I see that I need to instantiate a NoteParser now I need to answer a bunch of further questions to use responsibly: who should own the NoteParser? How long does it need to stay allocated? What are the implications of calling `parse(...)` on its internal state? Of course these aren't that hard to answer by reading the code, but it's worse API design because I don't need to read the code in the first example.

I know that sometimes people want something they can instantiate so they can mock certain functions for a test. In that case I think it's best to use a `class` and mark all the static functions `class func`, but we aren't doing any mocking for NoteParser so I don't think that applies.

(Also feel free to merge this if Josh approves. I do want to finish this discussion but I don't want it to hold this feature from shipping)

Reply:
I agree that we should be merging this if @joshuatbrown given that we are discussing the style here and whatever we decide won't impact the functionality. Josh, mind reviewing again?

Regarding the discussion, it's hard to separate what is good and wrong from what is just a matter of style. I don't think `static` functions are stricter, they could be modifying `static` properties; for me, `static` functions are more commonly used for constructing objects or small functions meant to be shared, and not for coding behavior or algorithms, these should be encapsulated in objects, that's how I learned it by reading the Design Patterns book. 

I think the doubts you have when instantiating a NoteParser are a bit unfair. For example, when I use a DateFormatter instance I don't need to answer complex questions like: who should own the DateFormatter? How long does it need to stay allocated? or What are the implications of calling a function on its internal state?


Also, the way we had this before was misleading. We had different static functions, but in reality, most of them were used internally by the other (the main one) and were meant to be modified together because the implementation was very specific to the problem we wanted to solve. I couldn't just reuse a function to solve something different. I think we should hide all this in an object. I agree it could be a `class` though, but I don't think `static` functions give us an edge, and make future modifications harder to make.

Reply:
I think I agree with a lot of what both of you are saying.

> I find it a best practice to make pure functions static because it is a clear signal to the developer: "this function doesn't use self"

💯 

> If you want to collect a bunch of related pure functions together it makes sense to do so in a stateless enum, because you can't instantiate a stateless enum.

💯 

> I think the clients of NoteParser should not care about if the functions inside use self or not.

💯 

> I know that sometimes people want something they can instantiate so they can mock certain functions for a test.

🖐️ That's me!

> In that case I think it's best to use a class and mark all the static functions class func, but we aren't doing any mocking for NoteParser so I don't think that applies.

🎯 This is the key problem for me. We aren't doing any mocking for NoteParser. In order to test anything that uses NoteParser, like Event+CoreDataClass, we'd be testing NoteParser along with it. I'd rather test Event+CoreDataClass in isolation; this makes it (a) easier write tests for it, since less code is under test, and (b) easier to identify bugs in it, since we can eliminated NoteParser as the potential source.

If we could find a way to both make our code as strict as possible _and_ make it easy to isolate our code for testing, I'd be all for that. If we can't, I always prioritize testability.